### PR TITLE
test: tests for #1435

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
tests for #1435 

Supporting https://github.com/salesforce/lwc/pull/1443
## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

### CODE REVIEW

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js [hunk 1]
```diff
@@ -157,3 +159,94 @@ it('should associate the component stack when the invocation throws', () => {
...
+        registerTemplate(tmpl);
+        class Parent extends LightningElement {
+            constructor() {
+                super();
+                this.callback = disconnectedCallback;
+                this.hideChild = false;
+                parentInstance = this;
+            }
+            render() {
+                return !this.hideChild ? tmpl : () => [];
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: @caridy I wasn't successful recreating the scenario described in https://github.com/salesforce/lwc/issues/1435#issuecomment-517407202
Updating the tracked field causes the following error message `Error: Assert Violation: markComponentAsDirty() for [object:vm undefined (3)] cannot be called during rendering of [object:vm Parent (2)].`

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js [hunk 2]
```diff
@@ -157,3 +159,94 @@ it('should associate the component stack when the invocation throws', () => {
...
     expect(error.wcStack).toBe('<x-disconnected-callback-throw>');
 });
+
+describe('disconnectedCallback for components with a custom render()', () => {
+    it('should invoke disconnectedCallback for children when parent is removed', () => {
+        let disconnectedCallbackInvoked = false;
+        function disconnectedCallback() {
+            disconnectedCallbackInvoked = true;
+        }
+        function tmpl($api, $cmp) {
```
Comment (caridy) [inline packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: this should be imported, should not be manually created because it will break once we make changes to compiler.

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js [hunk 3]
```diff
@@ -157,3 +159,94 @@ it('should associate the component stack when the invocation throws', () => {
...
+                            disconnect: $cmp.callback,
+                        },
+                        key: 2,
+                    },
+                    []
+                ),
+            ];
+        }
+        registerTemplate(tmpl);
+        class Parent extends LightningElement {
```
Comment (caridy) [inline packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: this should be a local component.

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js [hunk 4]
```diff
@@ -157,3 +159,94 @@ it('should associate the component stack when the invocation throws', () => {
     expect(error.message).toBe('throw in disconnected');
     expect(error.wcStack).toBe('<x-disconnected-callback-throw>');
 });
+
+describe('disconnectedCallback for components with a custom render()', () => {
+    it('should invoke disconnectedCallback for children when parent is removed', () => {
```
Comment (caridy) [inline packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: is this test related to the issue? seems like a regular test.

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js [hunk 5]
```diff
@@ -157,3 +159,94 @@ it('should associate the component stack when the invocation throws', () => {
...
+        expect(disconnectedCallbackInvoked).toBe(true);
+    });
+
+    xit('should invoke disconnectedCallback for children when parent switches template', () => {
+        let parentInstance;
+        let disconnectedCallbackInvoked = false;
+        function disconnectedCallback() {
+            disconnectedCallbackInvoked = true;
+        }
+        function tmpl($api, $cmp) {
```
Comment (caridy) [inline packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: this should be a local component with two templates.

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js (hunk 1 above)
Comment (caridy) [reply packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: we do not accept an empty template anymore... it must be a compiled template.

Comment (caridy) [reply packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: the switch of the tracked template can happen in the setter of the public api.

File: packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js (hunk 4 above)
Comment (ravijayaramappa) [reply packages/integration-karma/test/component/LightningElement.disconnectedCallback/index.spec.js]: Yes, this is a regular test for a component with a custom render function. Added it since there was no test for this scenario.

File: packages/@lwc/engine/src/framework/vm.ts [hunk 6]
```diff
@@ -412,7 +412,7 @@ function runDisconnectedCallback(vm: VM) {
         // it will be re-rendered because we are disconnecting the reactivity
         // linking, so mutations are not automatically reflected on the state
         // of disconnected components.
-        markComponentAsDirty(vm);
+        vm.isDirty = true;
```
Comment (ekashida) [inline packages/@lwc/engine/src/framework/vm.ts]: What was this change? Was it a bug fix?

Comment (ravijayaramappa) [reply packages/@lwc/engine/src/framework/vm.ts]: Yes. 
1. When a parent is being rehydrated, the old children in the parent's shadow tree are disconnected and marked as dirty(so that in the future if the child is reused, we know it has to be rehydrated).
2. During runDisconnectedCallback() invocation for the child, we know the child is not in a rendering phase but the parent is. 
3. To mark the child as dirty, we were calling markComponentAsDirty(), but that has an assertion to verify that we are not in a rendering phase(to avoid infinite loop).
4. So we short circuit that and directly mark the child's vm as dirty.
